### ISSUE TITLE
SPIR-V Environment Spec does not describe required accuracy for OpFMod and OpFRem

### ISSUE THREAD

**Issue by bashbaug** [KhronosGroup/OpenCL-Docs#859]
See related discussion: https://github.com/KhronosGroup/OpenCL-CTS/issues/1548

In short, the OpenCL SPIR-V environment spec currently does not describe the required accuracy for the **OpFMod** and **OpFRem** instructions (although it does describe the required accuracy for the `fmod` and `remainder` extended instructions).

OpenCL SPIR-V environment spec reference:

https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_Env.html#_ulp_values_for_math_instructions_full_profile

The spirv_new CTS tests are currently testing that **OpFMod** and **OpFRem** matches the behavior of these OpenCL C instructions, which may or may not be correct:

```c
#define spirv_frem(a, b) fmod(a, b)              
#define spirv_fmod(a, b) copysign(fmod(a,b),b)  
```

For Vulkan, the accuracy of **OpFMod** and **OpFRem** is derived from `trunc` and `floor`, which makes them cheap but potentially highly inaccurate:

https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#spirvenv-op-prec

> The OpFRem and OpFMod instructions use cheap approximations of remainder, and the error can be large due to the discontinuity in trunc() and floor(). This can produce mathematically unexpected results in some cases, such as FMod(x,x) computing x rather than 0, and can also cause the result to have a different sign than the infinitely precise result.

What do we want to do for OpenCL?

**Comment by karolherbst** [comment]
For somebody trying to implement OpenCL on native drivers but also on top of Vulkan, I think it might be worth saying that if nothing is specified otherwise it follows Vulkan rules.

dealing with `OpFDiv` is already a bit of an issues, but manageable without too much overhead.

I think it's a good idea that with `SPIR-V` you can opt into getting less precision using the existing opcodes. It might make sense though to fill the gaps by adding more OpenCL C builtins if such exist e.g. adding more `native_` variants which would map 1:1 to SPIR-V instructions.

**Comment by StuartDBrady** [comment]
Could you help explain a little about the layering with Vulkan, please?  Naïvely, I had previously assumed the translation of _OpFRem_ instructions in SPIR-V using the Kernel execution into SPIR-V using the GLCompute execution model would be similar to the translation of uses of _OpExtInst OpenCL fmod_, and that translation of _OpFMod_ would be implemented in terms of translation of _OpFRem_.  Does layering of OpenCL on Vulkan not work the way that I would have assumed?

Note in particular that in the SPIR-V/LLVM Translator, LLVM's _frem_ instruction is translated into _OpFRem_.  The LangRef states "This is the same output as a libm 'fmod' function".  Perhaps this is a bug in the SPIR-V/LLVM Translator, and we should be translating to _fmod_ in the OpenCL extended instruction set, instead?

**Comment by karolherbst** [comment]
> Could you help explain a little about the layering with Vulkan, please? Naïvely, I had previously assumed the translation of _OpFRem_ instructions in SPIR-V using the Kernel execution into SPIR-V using the GLCompute execution model would be similar to the translation of uses of _OpExtInst OpenCL fmod_, and that translation of _OpFMod_ would be implemented in terms of translation of _OpFRem_. Does layering of OpenCL on Vulkan not work the way that I would have assumed?
> 

That would make things just difficult. `OpFRem` should be the same on both sides. The OpenCL SPIR-V Env spec does not specify the precision of `OpFRem` at all, and imho it shouldn't have stricter rules than Vulkan here in the first place.

If one wants to have the precision of OpenCL C builtins, the `OpExtInst` ones exist for that reason. We shouldn't make it harder than it's supposed to be. If you want to write fast code in OpenCL having the option of using SPIR-V instructions directly to get more speed if the precision doesn't matter is I think something which we should keep, though I suspect some/most implementations are more strict here.

> Note in particular that in the SPIR-V/LLVM Translator, LLVM's _frem_ instruction is translated into _OpFRem_. The LangRef states "This is the same output as a libm 'fmod' function". Perhaps this is a bug in the SPIR-V/LLVM Translator, and we should be translating to _fmod_ in the OpenCL extended instruction set, instead?

yeah.. from a gut feeling I'd say it shall translate to the `OpExtInst` instead as the SPIR-V env spec doesn't give any guarantees whatsoever on  `OpFRem`. But that's kind of out of scope, because OpenCL C only has `fmod` here in the first place.

**Comment by b-sumner** [comment]
I was against introducing OpFRem, and still believe that it should be removed rather than producing completely wrong answers in certain cases.  Why can't those that want an approximation that only works sometimes just generate the sequence of SPIR-V ops that produce it?

**Comment by bashbaug** [comment]
We need to support OpFRem and OpFMod because these are standard SPIR-V instructions, so removing them is not an option.  We do have options how they should behave, though.

#### Option 1: Derive behavior from `trunc` and `floor`, similar to Vulkan.

Pros: Provides a faster option than the `fmod` and `remainder` extended instructions in cases when high accuracy is not required.  Provides uniformity with Vulkan, simplifying layered implementations.

Cons: Does not match the currently tested behavior and changing the required accuracy could break existing SPIR-V users.

#### Option 2: Match the `fmod` and `remainder` extended instructions (0 ULP).

Pros: Matches the currently tested behavior, less likelihood for confusion.

Cons: Lower performance in the cases where high accuracy is not required, potentially complicates layered implementations or SPIR-V generators targeting OpenCL and Vulkan.

#### Other options?

We could look into other options like allowing the derived behavior with "unsafe math optimizations" enabled, but we'd need to justify the added complexity.

**Comment by jekstrand** [comment]
I'm a bit torn. On the one hand, I want consistency. On the other hand, this is the libclc implementation of `fmod()`: https://github.com/llvm/llvm-project/blob/main/libclc/generic/lib/math/clc_fmod.cl#L30.  It's horrible.  If you've got a hardware instruction that does precise fmod, that's great. If you don't and need an emulation, getting 0 ULP is a giant pile of work that you don't want unless you *really* need the precision.  Back to the first hand, if you want `x - y * floor(x/y)`, maybe just type that?  I guess it is 3-4x the instructions if someone happens to have a FREM instruction but are people really fmod-bound?  That would be a weird app...

From a driver compiler point of view, I really want **OpFMod** and **OpFRem** to really be the same for Vulkan and OpenCL. Mesa currently tries very hard to not have special cases in our SPIR-V parser for the target API.  I'd rather those mean the same thing regardless of whether you're targeting Vulkan or OpenCL and have a separate thing for if you want the super-precise version for which we probably have to invoke libclc.  Yeah, ok, I think I convinced myself that these should match Vulkan. :grin:

**Comment by bashbaug** [comment]
Yeah, our implementation of the precise fmod is similar (we don't have a dedicated HW instruction).

In some ways the only reason we're having this discussion is because we currently have two different ways to perform similar operations - in this case a core SPIR-V instruction and an extended instruction - therefore we can (possibly) assign different behavior (accuracy requirements) to each operation.  Two interesting philosophical questions are:

1. Are we OK having two ways of performing a similar operation or should we try to have only a single way to perform an operation?
2. In cases where we have two ways of performing a similar operation, woudl we prefer them to behave the same, or is it OK if they behave differently?

If we want to match Vulkan I think we'll need at least some cases where there are two ways of performing a similar operation, because of the differences in behavior (specifically, accuracy requirements).

**Comment by jekstrand** [comment]
The consistency argument cuts both ways, unfortunately. Do we want the SPIR-V opcode to behave consistently across APIs or do we want the two things with the same name in OpenCL to behave the same in OpenCL?

**Comment by karolherbst** [comment]
I'm a bit skeptical of treating SPIR-V opcodes and OpenCL C builtins the same.

For the builtins we have `ExtInst` and that's good and helpful. The biggest problem in making SPIR-V and CLC builtins the same is, that some builtins might be implemented on top of the SPIR-V instruction having the same name.

Not having a good example here, but e.g. to satisfy OpenCL's requirement on `OpFDiv` we have to do some scaling around hw div instructions. Some builtins might actually be implemented in a similar way.

Granted, that's mostly a Mesa issue as our internal program IR is SPIR-V, which we also use for linking, so other runtimes might not care about that as much as we do, but at least for us, everything goes through SPIR-V one way or the other.

**Comment by bashbaug** [comment]
Discussed in the November 29th teleconference.  We're trending towards matching Vulkan for the SPIR-V `OpFMod` and `OpFRem`.  I will make a PR to clarify this in the OpenCL SPIR-V Environment Spec.

**Comment by StuartDBrady** [comment]
As noted in today's WG call, _FPFastMathMode_ decorations are always valid for _OpFRem_, but they are only valid for _fmod_ in the _OpenCL Extended Instruction Set_ when using a SPIR-V environment that explicitly allows it.  As we don't yet have a SPIR-V environment for OpenCL for SPIR-V 1.3 or above, we can (and IMO, we should) make sure that this is included in any such environment.

@bashbaug, any thoughts as to how we should best track this?  (Also, we should update the environment to allow ZeroExtend and SignExtend for images with SPIR-V 1.4 - and probably do a more general search for "missing before version 1._x_".)

**Comment by alycm** [comment]
On the call on 2023-01-17 there was a concern that SPIR-V LLVM translator could be less accurate than the Vulkan requirements and thus not meet the changes in #871.  That would need further investigation.
